File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutterobject/BaseNativeHandler.java
package com.valentingrigorean.arcgis_maps_flutter.flutterobject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class BaseNativeHandler<T> implements NativeHandler {

    private final T nativeHandler;
    private boolean isDisposed = false;
    private NativeMessageSink messageSink;

    protected BaseNativeHandler(T nativeHandler) {
        this.nativeHandler = nativeHandler;
    }

    public T getNativeHandler() {
        return nativeHandler;
    }

    @Override
    public final void dispose() {
        if (isDisposed) {
            return;
        }
        isDisposed = true;
        disposeInternal();
    }

    @Override
    public void setMessageSink(@Nullable NativeMessageSink messageSink) {
        this.messageSink = messageSink;
    }

    protected boolean isDisposed() {
        return isDisposed;
    }

    protected void disposeInternal() {

    }

    protected void sendMessage(@NonNull String method, @Nullable Object args) {
        if (messageSink != null && !isDisposed) {
            messageSink.send(method, args);
        }
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutterobject/BaseNativeHandler.kt
package com.valentingrigorean.arcgis_maps_flutter.flutterobject

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel

abstract class BaseNativeHandler<T> protected constructor(val nativeHandler: T) : NativeHandler {
    private var messageSink: NativeMessageSink? = null
    private val coroutineJob = Job()
    protected val scope = CoroutineScope(Dispatchers.Main + coroutineJob)

    protected var isDisposed = false
        private set

    override fun dispose() {
        if (isDisposed) {
            return
        }
        isDisposed = true
        coroutineJob.cancel()
    }

    override fun setMessageSink(messageSink: NativeMessageSink?) {
        this.messageSink = messageSink
    }

    protected fun sendMessage(method: String, args: Any?) {
        if (messageSink != null && !isDisposed) {
            messageSink!!.send(method, args)
        }
    }
}